drivers: spi: fix CS documentation

gmarull opened the pull request:
`spi_cs_control` `cs` field is no longer a pointer, but a struct member. Clarify it must be zero-initialized if not used. Remove confusing comment about the need to use `device_is_ready()` on `cs` field, `spi_is_ready_dt()` is the only thing users need to do.

gmarull committed:
drivers: spi: fix CS documentation

`spi_cs_control` `cs` field is no longer a pointer, but a struct member.
Clarify it must be zero-initialized if not used. Remove confusing
comment about the need to use `device_is_ready()` on `cs` field,
`spi_is_ready_dt()` is the only thing users need to do.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

RobertGalatNordic reviewed (approved):
lgtm

kartben commented:
Interestingly, I was just looking at some of the warnings that appear when setting EXTRACT_ALL to NO (https://github.com/zephyrproject-rtos/zephyr/issues/56855) and `spi_config` is one example of -- I think -- improperly documented structs. I don't think the fields should be documented using @param, but instead (this is without your fix from this very PR):

```c
/**
 * @brief SPI controller configuration structure
 *
 * @warning Most drivers use pointer comparison to determine whether a
 * passed configuration is different from one used in a previous
 * transaction.  Changes to fields in the structure may not be
 * detected.
 */
struct spi_config {
	/** 
	 * @brief bus frequency in Hertz.
	 */
	uint32_t		frequency;
#if defined(CONFIG_SPI_EXTENDED_MODES)
	uint32_t		operation;
	uint16_t		slave;
	uint16_t		_unused;
#else
	/** 
	 * @brief operation bit field.
	 * 
	 * It has the following parts:
	 *
	 *     operational mode    [ 0 ]       - master or slave.
	 *     mode                [ 1 : 3 ]   - Polarity, phase and loop mode.
	 *     transfer            [ 4 ]       - LSB or MSB first.
	 *     word_size           [ 5 : 10 ]  - Size of a data frame in bits.
	 *     duplex              [ 11 ]      - full/half duplex.
	 *     cs_hold             [ 12 ]      - Hold on the CS line if possible.
	 *     lock_on             [ 13 ]      - Keep resource locked for the caller.
	 *     cs_active_high      [ 14 ]      - Active high CS logic.
	 *     format              [ 15 ]      - Motorola or TI frame format (optional).
	 * if @kconfig{CONFIG_SPI_EXTENDED_MODES} is defined:
	 *     lines               [ 16 : 17 ] - MISO lines: Single/Dual/Quad/Octal.
	 *     reserved            [ 18 : 31 ] - reserved for future use.
	 */
	uint16_t		operation;
	/**
	 * @brief slave number from 0 to host controller slave limit.
	*/
	uint16_t		slave;
#endif /* CONFIG_SPI_EXTENDED_MODES */
	/**
	 * @brief valid pointer on a struct spi_cs_control is CS line is emulated 
	 * through a gpio line, or NULL otherwise.
	*/
	struct spi_cs_control cs;
};
```

Current docs: https://docs.zephyrproject.org/latest/doxygen/html/structspi__config.html
With the "fix": 

<img width="772" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/d857d147-6e80-418d-a8fe-1b4e86a3040a">

Thoughts?

gmarull commented:
@kartben you're right, let me push a new patch fixing this.

gmarull committed:
drivers: spi: add opaque type to encode SPI operation flags

Since flags can use either 16/32-bit depending on
CONFIG_SPI_EXTENDED_MODES, add a new opaque type that uses the correct
bit-width depending on that option. This allows us to simplify the
structure layout.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: spi: fix spi_config structure documentation

`@param` can't be used to document struct fields. Inline each field
documentation and improve it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on include/zephyr/drivers/spi.h:
```diff
@@ -269,42 +269,45 @@ struct spi_cs_control {
+ * @brief SPI controller configuration structure
+ */
+struct spi_config {
+	/** Bus frequency in Hertz. */
```
I think it's not needed, see https://builds.zephyrproject.io/zephyr/pr/60292/docs/doxygen/html/structspi__config.html

I'm +1 to enable autobrief.

fabiobaltieri reviewed (approved):
beep 🤖

kartben reviewed (approved):
Thanks!!
